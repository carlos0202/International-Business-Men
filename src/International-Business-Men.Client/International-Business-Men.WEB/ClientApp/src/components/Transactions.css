.button-hover:hover {
    cursor: pointer!important;
}


.card.card-body {
    margin-bottom: 65px;
}