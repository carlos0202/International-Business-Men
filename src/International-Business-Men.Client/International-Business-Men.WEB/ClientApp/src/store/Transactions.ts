import { Action, Reducer } from 'redux';
import { AppThunkAction } from './';
import { ResponseModel } from './CurrencyRates';

// -----------------
// STATE - This defines the type of data maintained in the Redux store.

export interface TransactionsState {
    isLoading: boolean;
    sku?: string;
    transactions: Transaction[];
}

export interface Transaction {
    sku: string;
    currency: string;
    amount: number;
}


// -----------------
// ACTIONS - These are serializable (hence replayable) descriptions of state transitions.
// They do not themselves have any side-effects; they just describe something that is going to happen.

interface RequestTransactionsAction {
    type: 'REQUEST_TRANSACTIONS';
    sku?: string;
}

interface ReceiveTransactionsAction {
    type: 'RECEIVE_TRANSACTIONS';
    sku?: string;
    transactions: Transaction[];
}

// Declare a 'discriminated union' type. This guarantees that all references to 'type' properties contain one of the
// declared type strings (and not any other arbitrary string).
type KnownAction = RequestTransactionsAction | ReceiveTransactionsAction;

// ----------------
// ACTION CREATORS - These are functions exposed to UI components that will trigger a state transition.
// They don't directly mutate state, but they can have external side-effects (such as loading data).

export const actionCreators = {
    requestTransactions: (sku: string): AppThunkAction<KnownAction> => (dispatch, getState) => {
        // Only load data if it's something we don't already have (and are not already loading)
        const appState = getState();
        console.log(sku, appState, appState.transactions);
        if (appState && appState.transactions && sku !== appState.transactions.sku) {
            let route = sku.length ? `/Transactions/${sku}` : '/Transactions';
            fetch(`${process.env.REACT_APP_API_URL}${route}`)
                .then(response => response.json() as Promise<ResponseModel<Transaction>>)
                .then(data => {
                    dispatch({ type: 'RECEIVE_TRANSACTIONS', transactions: data.result });
                });

            dispatch({ type: 'REQUEST_TRANSACTIONS' });
        }
    }
};

// ----------------
// REDUCER - For a given state and action, returns the new state. To support time travel, this must not mutate the old state.

const unloadedState: TransactionsState = { transactions: [], isLoading: false };

export const reducer: Reducer<TransactionsState> = (state: TransactionsState | undefined, incomingAction: Action): TransactionsState => {
    if (state === undefined) {
        return unloadedState;
    }

    const action = incomingAction as KnownAction;
    switch (action.type) {
        case 'REQUEST_TRANSACTIONS':
            return {
                transactions: state.transactions,
                isLoading: true
            };
        case 'RECEIVE_TRANSACTIONS':
            if (action.transactions) {
                return {
                    transactions: action.transactions,
                    isLoading: false
                };
            }
    }

    return state;
};
