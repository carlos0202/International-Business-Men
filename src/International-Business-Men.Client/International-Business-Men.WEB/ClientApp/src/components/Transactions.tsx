import * as React from 'react';
import { connect } from 'react-redux';
import { ApplicationState } from '../store';
import * as TransactionsStore from '../store/Transactions';
import * as CurrencyRatesStore from '../store/CurrencyRates';
import { RouteComponentProps } from 'react-router';
import {
    Form, InputGroup,
    InputGroupAddon,
    Input,
    Button,
    Row,
    Col,
    Card,
    CardTitle
} from 'reactstrap';
import { bindActionCreators } from 'redux';
import TransactionsTable from './TransactionsTable';
import {
    getSupportedCurrencies
} from '../utils/transformations';
import './Transactions.css';

// At runtime, Redux will merge together...
type TransactionProps =
    TransactionsStore.TransactionsState
    & CurrencyRatesStore.CurrencyRatesState// ... state we've requested from the Redux store
    & typeof TransactionsStore.actionCreators // ... plus action creators we've requested
    & typeof CurrencyRatesStore.actionCreators
    & RouteComponentProps<{ sku: string }>;

class Transactions extends React.PureComponent<TransactionProps, {}, { sku: string, convertCurrency: string }> {
    public state = {
        sku: this.props.match.params.sku || "",
        convertCurrency: 'EUR'
    };

    // This method is called when the component is first added to the document
    public componentDidMount() {
        this.ensureDataFetched();
    }

    onChange = (evt: any) => this.setState({ sku: evt.target.value });

    onCurrencyChange = (evt: any) => this.setState({ convertCurrency: evt.target.value });

    onClear = (evt: any) => {
        evt.preventDefault();
        this.setState({ sku: "" });
        this.props.history.push(`/Transactions`);
        this.props.requestCurrencyRates();
        this.props.requestTransactions("");
    }

    onSubmit = (e: any) => {
        e.preventDefault();
        let sku = this.state.sku ? this.state.sku : "";

        if (!sku) return;

        this.props.requestCurrencyRates();
        this.props.history.push(`/Transactions/${sku}`);
        this.props.requestTransactions(sku);
    }

    public render() {

        return (
            <Card body>
                <CardTitle><h1>Transacciones</h1></CardTitle>
                <Row>
                    <Col md="9">
                        <Form onSubmit={this.onSubmit.bind(this)}>
                            <InputGroup>
                                <InputGroupAddon addonType="prepend" onClick={this.onClear.bind(this)} className="button-hover">
                                    <Button className="bg-default">Limpiar</Button>
                                </InputGroupAddon>
                                <Input type="text" id="sku" onChange={this.onChange.bind(this)} placeholder="SKU del Producto"
                                    value={this.state.sku} />
                                <InputGroupAddon addonType="append" onClick={this.onSubmit.bind(this)} className="button-hover">
                                    <Button className=" bg-primary">Buscar</Button>
                                </InputGroupAddon>
                            </InputGroup>
                        </Form>
                    </Col>
                    <Col md="3">
                        <InputGroup>
                            <Input type="select" name="convertCurrency" id="convertCurrency" onChange={this.onCurrencyChange.bind(this)}
                                value={this.state.convertCurrency} placeholder="Moneda de Conversión">
                                {Array.from(getSupportedCurrencies(this.props.currencyRates)).map(e => <option key={e} value={e}>{e}</option>)}
                            </Input>
                        </InputGroup>
                    </Col>
                </Row>

                <br />
                <br />
                <TransactionsTable currencyRates={this.props.currencyRates} convertCurrency={this.state.convertCurrency}
                    transactions={this.props.transactions} isLoading={this.props.isLoading} currentPage={0} />
            </Card>
        );
    }

    private ensureDataFetched() {
        this.props.requestCurrencyRates();
        const sku = this.props.match.params.sku || "";
        this.props.requestTransactions(sku);
    }
}

export default connect(
    (state: ApplicationState) => {
        const { transactions, currencyRates } = state;
        const isLoading = (transactions && transactions.isLoading) ||
            (currencyRates && currencyRates.isLoading);

        return { ...transactions, ...currencyRates, isLoading };
    }, // Selects which state properties are merged into the component's props
    (dispatch: any) => {
        return {
            ...bindActionCreators(Object.assign({}, TransactionsStore.actionCreators, CurrencyRatesStore.actionCreators), dispatch)
        };
    }
    //TransactionsStore.actionCreators // Selects which action creators are merged into the component's props
)(Transactions as any);
