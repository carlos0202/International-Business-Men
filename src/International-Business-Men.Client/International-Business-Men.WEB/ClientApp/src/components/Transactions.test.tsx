import * as React from 'react';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import { Transactions } from './Transactions';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { currencyRates, transactions } from '../../test/setup/test-data.json';
import thunk from 'redux-thunk';
import { createMemoryHistory, createLocation } from 'history';
import { match } from 'react-router';

const history = createMemoryHistory();
const path = `/Transactions/:sku`;

const matcher: match<{ sku: string }> = {
    isExact: false,
    path,
    url: path.replace(':sku', '1'),
    params: { sku: "1" }
};

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const location = createLocation(matcher.url);

describe('', () => {
    let store;
    let component;
    let isLoading = false;

    beforeEach(() => {
        store = mockStore({
            currencyRates, transactions, isLoading
        });
        component = renderer.create(
            <Provider store={store}>
                <MemoryRouter>
                    <Transactions {...store} currencyRates={currencyRates} transactions={transactions}
                        isLoading={isLoading} requestCurrencyRates={jest.fn()} requestTransactions={jest.fn("")}
                        history={history} location={location} match={matcher}  />
                </MemoryRouter>                
            </Provider>
        );
    });

    it('renders Transactions component witouth crashing and to match snapshot', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });
});

