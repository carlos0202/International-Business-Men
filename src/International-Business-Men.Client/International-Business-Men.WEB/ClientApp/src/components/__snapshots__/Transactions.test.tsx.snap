// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Transactions component witouth crashing and to match snapshot 1`] = `
<div
  className="card card-body"
>
  <div
    className="card-title"
  >
    <h1>
      Transacciones
    </h1>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-9"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="input-group"
        >
          <div
            className="button-hover input-group-prepend"
            onClick={[Function]}
          >
            <button
              aria-label={null}
              className="bg-default btn btn-secondary"
              onClick={[Function]}
            >
              Limpiar
            </button>
          </div>
          <input
            className="form-control"
            id="sku"
            onChange={[Function]}
            placeholder="SKU del Producto"
            type="text"
            value="1"
          />
          <div
            className="button-hover input-group-append"
            onClick={[Function]}
          >
            <button
              aria-label={null}
              className=" bg-primary btn btn-secondary"
              onClick={[Function]}
            >
              Buscar
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-md-3"
    >
      <div
        className="input-group"
      >
        <select
          className="form-control"
          id="convertCurrency"
          name="convertCurrency"
          onChange={[Function]}
          placeholder="Moneda de Conversi�n"
          value="EUR"
        >
          <option
            value="EUR"
          >
            EUR
          </option>
          <option
            value="USD"
          >
            USD
          </option>
          <option
            value="CAD"
          >
            CAD
          </option>
          <option
            value="AUD"
          >
            AUD
          </option>
        </select>
      </div>
    </div>
  </div>
  <br />
  <br />
  <div
    className="row"
  >
    <div
      className="col-xl-7 d-flex justify-content-start"
    >
      <h4>
        Total 
        <span
          className="text-muted"
        >
          5
        </span>
         transacciones: 
        <span
          className="text-primary"
        >
          EUR
        </span>
         
        65.05
      </h4>
    </div>
    <div
      className="col-xl-3 d-flex justify-content-center"
    >
      <select
        className="form-control"
        id="convertCurrency"
        name="pageSize"
        onChange={[Function]}
        placeholder="Registros por página"
        value={10}
      >
        <option
          value={10}
        >
          10
        </option>
        <option
          value={50}
        >
          50
        </option>
        <option
          value={100}
        >
          100
        </option>
        <option
          value={5}
        >
          Todos
        </option>
      </select>
    </div>
    <div
      className="col-xl-2 d-flex justify-content-lg-end justify-content-center"
    >
      <nav
        aria-label="Transactions table pagination"
        className=""
      >
        <ul
          className="pagination"
        >
          <li
            className="page-item disabled"
          >
            <a
              aria-label="First"
              className="page-link"
              href="#"
              onClick={[Function]}
            >
              <span
                aria-hidden="true"
              >
                «
              </span>
              <span
                className="sr-only"
              >
                First
              </span>
            </a>
          </li>
          <li
            className="page-item disabled"
          >
            <a
              aria-label="Previous"
              className="page-link"
              href="#"
              onClick={[Function]}
            >
              <span
                aria-hidden="true"
              >
                ‹
              </span>
              <span
                className="sr-only"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            className="page-item active"
          >
            <a
              className="page-link"
              href="#"
              onClick={[Function]}
            >
              1
            </a>
          </li>
          <li
            className="page-item disabled"
          >
            <a
              aria-label="Next"
              className="page-link"
              href="#"
              onClick={[Function]}
            >
              <span
                aria-hidden="true"
              >
                ›
              </span>
              <span
                className="sr-only"
              >
                Next
              </span>
            </a>
          </li>
          <li
            className="page-item disabled"
          >
            <a
              aria-label="Last"
              className="page-link"
              href="#"
              onClick={[Function]}
            >
              <span
                aria-hidden="true"
              >
                »
              </span>
              <span
                className="sr-only"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
  <div
    className="table-responsive-md"
  >
    <table
      aria-labelledby="tableLabel"
      className="table table-striped table-bordered table-hover"
    >
      <thead
        className="thead-dark"
      >
        <tr>
          <th>
            SKU del Producto
          </th>
          <th>
            Moneda
          </th>
          <th>
            Monto Transacción
          </th>
          <th>
            Monto Convertido
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            T2006
          </td>
          <td>
            USD
          </td>
          <td>
            USD
             
            10
          </td>
          <td>
            EUR
             
            7.36
          </td>
        </tr>
        <tr>
          <td>
            M2007
          </td>
          <td>
            CAD
          </td>
          <td>
            CAD
             
            34.57
          </td>
          <td>
            EUR
             
            25.31
          </td>
        </tr>
        <tr>
          <td>
            R2008
          </td>
          <td>
            AUD
          </td>
          <td>
            AUD
             
            17.95
          </td>
          <td>
            EUR
             
            9.12
          </td>
        </tr>
        <tr>
          <td>
            T2006
          </td>
          <td>
            EUR
          </td>
          <td>
            EUR
             
            7.63
          </td>
          <td>
            EUR
             
            7.63
          </td>
        </tr>
        <tr>
          <td>
            B2009
          </td>
          <td>
            USD
          </td>
          <td>
            USD
             
            21.23
          </td>
          <td>
            EUR
             
            15.63
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
